src/ripple/ledger/impl/ApplyView.cpp

```diff
@@ -0,0 +1,298 @@
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <BeastConfig.h>
+#include <ripple/ledger/ApplyView.h>
+#include <ripple/basics/contract.h>
+#include <ripple/protocol/Protocol.h>
+#include <cassert>
+
+namespace ripple {
+
+void
+ApplyView::creditHook (
+    AccountID const& from,
+    AccountID const& to,
+    STAmount const& amount,
+    STAmount const& preCreditBalance)
+{
+}
+
+// Called when the owner count changes
+// This is required to support PaymentSandbox
+void
+ApplyView::adjustOwnerCountHook (
+    AccountID const& account,
+    std::uint32_t cur,
+    std::uint32_t next)
+{
+
+}
+
+std::pair<bool, std::uint64_t>
+ApplyView::dirInsert (
+    Keylet const& directory,
+    uint256 const& key,
+    bool strictOrder,
+    std::function<void (std::shared_ptr<SLE> const&)> describe)
+{
+    assert(keylet::page(directory, 0).key == directory.key);
+
+    auto root = peek(directory);
+
+    if (! root)
+    {
+        // No root, make it.
+        root = std::make_shared<SLE>(directory);
+        root->setFieldH256 (sfRootIndex, directory.key);
+        describe (root);
+
+        STVector256 v;
+        v.push_back (key);
+        root->setFieldV256 (sfIndexes, v);
+
+        insert (root);
+
+        return { true, 0 };
+    }
+
+    std::uint64_t page = root->getFieldU64(sfIndexPrevious);
+
+    auto node = root;
+
+    if (page)
+    {
+        node = peek (keylet::page(directory, page));
+
+        if (!node)
+            LogicError ("Directory chain: root back-pointer broken.");
+    }
+
+    auto indexes = node->getFieldV256(sfIndexes);
+
+    // If there's space, we use it:
+    if (indexes.size () < dirNodeMaxEntries)
+    {
+        // If we're not maintaining strict order, then
+        // sort entries lexicographically.
+        if (!strictOrder)
+        {
+            auto v = indexes.value();
+
+            // We can't use std::lower_bound here because
+            // existing pages may not be sorted
+            auto pos = std::find_if(v.begin(), v.end(),
+                [&key](uint256 const& item)
+                {
+                    return key < item;
+                });
+
+            if (pos != v.end() && key == *pos)
+                LogicError ("dirInsert: double insertion");
+
+            v.insert (pos, key);
+            indexes = v;
```

Comment:
We can save two copies of the vector if we add an `insert`member to `STVector`. We'd save the copy on line 96 (it would becore `auto const& v = ...`), and the copy here. Since `STVector` already has an `erase` member `insert` seems OK.